doc/reference/kernel/scheduling/index.rst

```diff
@@ -31,6 +31,74 @@ exist, the scheduler chooses the one that has been waiting longest.
 
+
+The kernel can be built with with several choices for the
+ready queue implementation, offering different choices between
+code size, constant factor runtime overhead and performance
+scaling when many threads are added.
+
+* Simple linked-list ready queue
+
+  The scheduler ready queue will be implemented as a simple unordered list, with
+  very fast constant time performance for single threads and very low code size.
+  This implementation should be selected on systems with constrained code size
+  that will never see more than a small number (3, maybe) of runnable threads in
+  the queue at any given time.  On most platforms (that are not otherwise using
+  the red/black tree) this results in a savings of ~2k of code size.
+
+* Red/black tree ready queue"
+
+  The scheduler ready queue will be implemented as a red/black tree.  This has
+  rather slower constant-time insertion and removal overhead, and on most
+  platforms (that are not otherwise using the rbtree somewhere) requires an
+  extra ~2kb of code.  But the resulting behavior will scale cleanly and
+  quickly into the many thousands of threads.
+
+  Use this on platforms where you may have many threads (very roughly: more
+  than 20 or so) marked as runnable at a given time.
+
+  .. note::
+
+      Most applications don't want this.
+
+* Traditional multi-queue ready queue
+
+  When selected, the scheduler ready queue will be implemented as the
+  classic/textbook array of lists, one per priority (max 32 priorities).
+
+  This corresponds to the scheduler algorithm used in Zephyr versions prior to
+  1.12.
+
+  It incurs only a tiny code size overhead vs. the "dumb" scheduler and runs in
+  O(1) time in almost all circumstances with very low constant factor.  But it
+  requires a fairly large RAM budget to store those list heads, and the limited
+  features make it incompatible with features like deadline scheduling that
+  need to sort threads more finely, and SMP affinity which need to traverse the
+  list of threads.
+
+  Typical applications with small numbers of runnable threads probably want the
+  DUMB scheduler.
+
+
+The wait_q abstraction used in IPC primitives to pend threads for later wakeup
+shares the same backend data structure choices as the scheduler, and can use
+the same options.
+
+* Scalable wait_q implementation
+
+  When selected, the wait_q will be implemented with a balanced tree.  Choose
+  this if you expect to have many threads waiting on individual primitives.
+  There is a ~2kb code size increase over WAITQ_DUMB (which may be shared with
+  SCHED_SCALABLE) if the rbtree is not used elsewhere in the application, and
+  pend/unpend operations on "small" queues will be somewhat slower (though this
+  is not generally a performance path).
+
+* Simple linked-list wait_q
+
+  When selected, the wait_q will be implemented with a doubly-linked list.
+  Choose this if you expect to have only a few threads blocked on any single
+  IPC primitive.
+
```

Comment:
There are more SCHED options not mentioned here: ``CONFIG_SCHED_DEADLINE``, ``CONFIG_SCHED_CPU_MASK``, ``CONFIG_SCHED_IPI_SUPPORTED``, and ``CONFIG_SCHED_IPI_VECTOR``

Reply:
lets start with the above for now..